src/main/java/org/opensearch/security/resources/ResourceAccessHandler.java

```diff
@@ -0,0 +1,401 @@
+        ).collect(Collectors.toSet());
+
+        BoolQueryBuilder query = QueryBuilders.boolQuery()
+            .should(QueryBuilders.termQuery("created_by.user.keyword", user.getName()))
+            .should(QueryBuilders.termsQuery("all_shared_principals", flatPrincipals))
+            .minimumShouldMatch(1);
+
+        // 3) Fetch all accessible resource IDs
+        resourceSharingIndexHandler.fetchSharedDocuments(resourceIndex, flatPrincipals, query, listener);
+    }
+
+    /**
+     * Checks whether current user has permission to access given resource.
+     *
+     * @param resourceId    The resource ID to check access for.
+     * @param resourceIndex The resource index containing the resource.
+     * @param actionGroups  The set of action groups to check permission for.
+     * @param listener      The listener to be notified with the permission check result.
+     */
+    public void hasPermission(String resourceId, String resourceIndex, Set<String> actionGroups, ActionListener<Boolean> listener) {
+        validateArguments(resourceId, resourceIndex, actionGroups);
+
+        final UserSubjectImpl userSubject = (UserSubjectImpl) threadContext.getPersistent(
+            ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER
+        );
+        final User user = (userSubject == null) ? null : userSubject.getUser();
+
+        if (user == null) {
+            LOGGER.warn("No authenticated user found. Access to resource {} is not authorized.", resourceId);
+            listener.onResponse(false);
+            return;
+        }
+
+        LOGGER.info("Checking if user '{}' has permission to resource '{}'", user.getName(), resourceId);
+
+        if (adminDNs.isAdmin(user)) {
+            LOGGER.debug("User '{}' is admin, automatically granted permission on '{}'", user.getName(), resourceId);
+            listener.onResponse(true);
+            return;
+        }
+
+        Set<String> userRoles = new HashSet<>(user.getSecurityRoles());
+        Set<String> userBackendRoles = new HashSet<>(user.getRoles());
+
+        this.resourceSharingIndexHandler.fetchResourceSharingDocument(resourceIndex, resourceId, ActionListener.wrap(document -> {
+            if (document == null) {
+                LOGGER.warn(
+                    "ResourceSharing entry not found for '{}' and index '{}'. Access to this resource will be allowed.",
+                    resourceId,
+                    resourceIndex
+                );
+                // Since no sharing entry exists, requests is allowed to implement a non-breaking behaviour
+                listener.onResponse(true);
+                return;
+            }
+
+            // All public entities are designated with "*"
+            userRoles.add("*");
+            userBackendRoles.add("*");
+            if (isOwnerOfResource(document, user.getName())
+                || isSharedWithEveryone(document)
```

Comment:
Why not have this as methods within `ResourceSharing`? i.e. `document.isSharedWithEveryone()`.

Similarly wdyt about other methods like `document.isSharedWithUser()`, `document.isSharedWithRole()`, etc. You can then abstract the handling for the `*` case where the document is public. 

I don't think you need separate handling for `*` because `isSharedWithEveryone` should already return true by the time you evaluate the next conditions in this clause.

Reply:
Are you suggesting we change the design of the document?

This computation is fairly quick since it will only be evaluating for 1 resource at any time. We should not trade this to introduces new pre-computations.

OR are you suggesting we simply move this computation to ResourceSharing.java?

Reply:
I've moved them to ResourceSharing.java